spanstat: Report asymmetric use of Start() and End() in developer builds

tgraf opened the pull request:
This code has not been released yet so no backports required

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5478)
<!-- Reviewable:end -->

tgraf committed:
spanstat: Ignore End() without Start()

Handle the developer mistake of End() without Start()

Fixes: 89cbe1f56a9 ("endpoint: Rework regeneration logging with detailed duration times")

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
stackdump: Factor stack frame printing into pkg/stackdump

It also ensures that the error message is printed to the same io.Writer

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
spanstat: Print stack frame if End() is called without Start()

Example of unit test:
```
goroutine 20 End() called without prior Start()
goroutine 20 [running]:
goroutine 20 runtime/debug.Stack(0xc4200664c0, 0x591e20, 0xc4200664c0)
goroutine 20 	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
goroutine 20 github.com/cilium/cilium/pkg/stackdump.Errorf(0x5bf8e7, 0x22, 0x0, 0x0, 0x0)
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/stackdump/stackdump.go:26 +0x26
goroutine 20 github.com/cilium/cilium/pkg/spanstat.(*SpanStat).End(0xc42004cb30)
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/spanstat/spanstat.go:40 +0xc0
goroutine 20 github.com/cilium/cilium/pkg/spanstat.(*SpanStatTestSuite).TestSpanStat(0x6ce9c0, 0xc4200a21e0)
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/spanstat/spanstat_test.go:39 +0xfa
goroutine 20 reflect.Value.call(0x583720, 0x6ce9c0, 0x213, 0x5b8b53, 0x4, 0xc42004cf70, 0x1, 0x1, 0x6cf860, 0x5b2e80, ...)
goroutine 20 	/usr/local/go/src/reflect/value.go:447 +0x969
goroutine 20 reflect.Value.Call(0x583720, 0x6ce9c0, 0x213, 0xc42002cf70, 0x1, 0x1, 0x0, 0xc42002cf07, 0x0)
goroutine 20 	/usr/local/go/src/reflect/value.go:308 +0xa4
goroutine 20 github.com/cilium/cilium/vendor/gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc4200a21e0)
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:772 +0x575
goroutine 20 github.com/cilium/cilium/vendor/gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc4200a0080, 0xc4200a21e0, 0xc420080180)
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:666 +0x7c
goroutine 20 created by github.com/cilium/cilium/vendor/gopkg.in/check%2ev1.(*suiteRunner).forkCall
goroutine 20 	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:663 +0x20c
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on pkg/spanstat/spanstat.go:
```diff
@@ -32,7 +34,13 @@ func (s *SpanStat) Start() {
+	if !s.spanStart.IsZero() {
+		s.totalDuration += time.Since(s.spanStart)
+	} else {
+		stackdump.Errorf("End() called without prior Start()")
```
What can we possible gain by printing the errorf with a long stack trace? If this is a developer's mistake why not panic instead?

tgraf replied:
I think the only plausible justification for a panic is if the system is inconsistent and continued operation is not possible. If we panic because of harmless errors like these, we are just causing downtime in systems for no good reason. I'd rather scan logs for these errors than have users yell at us because we panic for a minor bug like this.

aanm replied:
But we should caught the panics in our CI. Also, golang itself has lots of panics built-in for situations like these. For example, if you call a `WaitGroup.Add()` after a `WaitGroup.Done()` you get a panic. https://play.golang.org/p/5OeXcShdnev

tgraf replied:
An `.Add()` after a `.Done()` can potentially cause a deadlock and may justify this. At the same time, the golang runtime does not have a standardized way of reporting such problems without introducing error handling into every API function.

I really don't buy into the value of crashing the agent and cause a disruption in all L7 forwarding for simple developer mistakes. I'm happy to be convinced otherwise but we need more justification than just "others are doing similar".

raybejjani replied:
since stackdump is a package, we could use build flags to toggle whether we panic or just print (similar to our lockedebug toggle). The CI does scan for specific strings (like panic), though, so it might be as effective to just add a fixed start string that the CI checks for.

tgraf replied:
@raybejjani I like that idea a lot! Will add this.

tgraf replied:
Complication. This will now no longer allow to actually test the spanstat package to catch these conditions without causing panics.

tgraf replied:
I guess it is possible to to keep the build tag out of the `go test` invocation and only pass it into the build.

raybejjani replied:
If you add the string to https://github.com/cilium/cilium/blob/master/test/helpers/cons.go#L192-L203 the CI will also catch this (not quite the panic aanm wants but it is still developer feedback).

aanm reviewed (dismissed):
I'm not sure if this PR is supposed to be reviewed yet but I've left my 2 cents.

tgraf committed:
build: Add -tags debug if DEBUG=1 is set

Similiar to LOCKDEBUG=1, introduce DEBUG=1 as generic debug build flag and
enable it for runtime tests.

Signed-off-by: Thomas Graf <thomas@cilium.io>
